**Daily Challenge: Building Your First Neural Network on the MNIST Dataset:**

In [3]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense
from tensorflow.keras.utils import to_categorical

from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay, classification_report

print("TensorFlow version:", tf.__version__)

TensorFlow version: 2.19.0


In [5]:
# Load data
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [6]:
# Normalize
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

In [7]:
# One-hot encode
y_train_oh = to_categorical(y_train, 10)
y_test_oh = to_categorical(y_test, 10)

In [12]:
# Model
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(32, activation="relu"),
    Dense(16, activation="relu"),
    Dense(10, activation="softmax")
])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [13]:
# Compile
model.compile(
    optimizer="adam",
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)

In [14]:
# Train
history = model.fit(
    x_train,
    y_train_oh,
    epochs=5,
    batch_size=32,
    validation_split=0.1
)

Epoch 1/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8864 - loss: 0.3901 - val_accuracy: 0.9543 - val_loss: 0.1702
Epoch 2/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9447 - loss: 0.1894 - val_accuracy: 0.9613 - val_loss: 0.1444
Epoch 3/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9546 - loss: 0.1548 - val_accuracy: 0.9635 - val_loss: 0.1319
Epoch 4/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9605 - loss: 0.1346 - val_accuracy: 0.9653 - val_loss: 0.1274
Epoch 5/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9637 - loss: 0.1191 - val_accuracy: 0.9675 - val_loss: 0.1250


In [15]:
# Evaluate
test_loss, test_acc = model.evaluate(x_test, y_test_oh)

print("Test loss:", test_loss)
print("Test accuracy:", test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9611 - loss: 0.1316
Test loss: 0.13156279921531677
Test accuracy: 0.9610999822616577


I implemented a simple fully connected neural network with two small hidden layers (32 and 16 neurons) using ReLU activation. The output layer uses softmax to produce probabilities over the 10 digit classes. Test loss is 0.13 and Test accuracy is 0.96.